process/process.go

```diff
@@ -304,10 +311,96 @@ func killWindowsProcessTreeRecursive(childProcess *psutil.Process) {
+	v    uint64
+	lock sync.Mutex
+}
+
+func (c *SafeCounter) readThenInc() uint64 {
+	c.lock.Lock()
+	value := c.v
+	c.v++
+	c.lock.Unlock()
+	return value
+}
+
+type TopLevelEnum int
+
+const (
+	NotTopLevel TopLevelEnum = 0
+	TopLevel    TopLevelEnum = 1
+)
+
+// lexicographically encode integer
+// based on http://www.zanopha.com/docs/elen.pdf
+func lexEncode(n uint64, topLevel TopLevelEnum) string {
+	var encoded string
+	// recursively calculate lex prefix
+	// the lex prefix allows the user to lexicographically sort the output
+	// need lex prefix if n has more than one digit
+	nstr := fmt.Sprintf("%d", n)
+	nlen := uint64(len(nstr))
+	if nlen > 1 {
+		// include non-numeric part of lex prefix
+		// to allow proper sorting, this char must come after numerics in the ascii table
+		encoded = fmt.Sprintf("_")
+		// then include recursive part
+		encoded += lexEncode(nlen, NotTopLevel)
+		// conditionally include lex separator
+		if topLevel == TopLevel {
+			// the lex separator allows the user to differentiate a numeric part of the lex prefix from the original number
+			// to allow proper sorting, the lex separator must come before numerics in the ascii table
+			encoded += "."
+		}
+	}
+	// include numeric part of lex prefix, or
+	// original number (if topLevel==true)
+	encoded += nstr
+	return encoded
+}
+
+func getEntrySeparator() string {
+	// to allow proper sorting, the entry separator must come before numerics in the ascii table
+	return "/"
+}
+
+func writeImmediateLines(numJobs int, cmdId uint64, tryNumber int, lineNumber *SafeCounter, reader *bufio.Reader, fh *os.File) error {
+	var err error = nil
+	for {
+		// read line by line
+		var line string
+		if reader != nil {
+			line, err = reader.ReadString('\n')
+			// only write non-empty lines
+			if len(line) > 0 && line != "\n" && line != "\r\n" {
+				// we only need to prefix output if running in parallel
+				if numJobs > 1 {
+					// we prefix lines so that the user can differentiate the contents of the interleaved output
+					// use parentheses, encoded integers, and a colon, so that the user can afterward sort the output lines
+					// and get the output correctly grouped by command id, try number, and line
+					lineNumberValue := lineNumber.readThenInc()
+					prefix := fmt.Sprintf("(%s", lexEncode(cmdId, TopLevel))
+					prefix += fmt.Sprintf("%s%s", getEntrySeparator(), lexEncode(uint64(tryNumber), TopLevel))
+					prefix += fmt.Sprintf("%s%s): ", getEntrySeparator(), lexEncode(lineNumberValue, TopLevel))
+					line = prefix + line
+				}
+				fh.WriteString(line)
+			}
+		} else {
+			err = io.EOF
+		}
+		if err != nil {
+			break
+		}
+	}
+	return err
+}
+
 // run a command and pass output to c.reader.
 // Note that output returns only after finishing run.
 // This function is mainly borrowed from https://github.com/brentp/gargs .
-func (c *Command) run(opts *Options) error {
+func (c *Command) run(opts *Options, tryNumber int, outfh *os.File, errfh *os.File) error {
```

Comment:
Can you move the `outfh` and `errfh` to `Options`, that's feasible and would makes it more elegant.